File: services/atlas-terminal/test/conformance.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it } from 'bun:test';

describe('cross-surface reference missions', () => {
	it('keeps atlas-terminal on the shared gateway event contract', () => {
		const fixture = JSON.parse(
			readFileSync(resolve(process.cwd(), '../agent-runtime/tests/fixtures/reference_missions.json'), 'utf8')
		) as { missions: Array<{ id: string; surface_projections: Record<string, Array<{ event_index: number; kind: string }>> }> };

		expect(fixture.missions).toHaveLength(8);
		for (const mission of fixture.missions) {
			const events = mission.surface_projections.atlas_terminal;
			expect(events.length, `${mission.id}: terminal projection`).toBeGreaterThan(0);
			expect(events.map((event) => event.event_index), `${mission.id}: ordered events`).toEqual(events.map((_, index) => index));
		}
	});
});

File: services/atlas-tui/internal/client/conformance_test.go
package client

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func TestCrossSurfaceReferenceMissionsUseGatewayClientContract(t *testing.T) {
	path := filepath.Join("..", "..", "..", "agent-runtime", "tests", "fixtures", "reference_missions.json")
	body, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Surfaces []string `json:"surfaces"`
		Missions []struct {
			ID                 string `json:"id"`
			SurfaceProjections map[string][]struct {
				EventIndex int    `json:"event_index"`
				Kind       string `json:"kind"`
			} `json:"surface_projections"`
		} `json:"missions"`
	}
	if err := json.Unmarshal(body, &fixture); err != nil {
		t.Fatal(err)
	}
	if len(fixture.Missions) != 8 {
		t.Fatalf("reference mission count = %d, want 8", len(fixture.Missions))
	}
	for _, mission := range fixture.Missions {
		projection := mission.SurfaceProjections["go_tui"]
		if len(projection) == 0 {
			t.Fatalf("%s has no Go TUI gateway projection", mission.ID)
		}
		for index, event := range projection {
			if event.EventIndex != index {
				t.Fatalf("%s event[%d].event_index = %d, want ordered gateway event", mission.ID, index, event.EventIndex)
			}
		}
	}
}
